### ISSUE TITLE
FSDP: unsharded grads accessor

### ISSUE THREAD

**Issue by stas00** [pytorch/pytorch#120879]
### 🚀 The feature, motivation and pitch

moving from [slack discussion](https://pytorch.slack.com/archives/C3PDTEV8E/p1709153741504309) 

How do I access full unsharded gradients under FSDP (Deepspeed ZeRO has a special API to gather grads on demand since it's expensive)
Andrew Gu replied:

>  1. The support is not great. I added a `with_grads=True` option to `summon_full_params()` specifically  for `use_orig_params=True` for debugging. There might be some rough edges there.

### Alternatives

Deepspeed ZeRO provides accessors for unsharded grads and optim states - the full doc is [here](https://deepspeed.readthedocs.io/en/latest/zero3.html#debugging)

2. deepspeed.utils.safe_get_full_fp32_param(param)
3. deepspeed.utils.safe_get_full_grad(param)
4. deepspeed.utils.safe_get_full_optimizer_state(param, optim_state_key)
5. deepspeed.utils.safe_get_local_fp32_param(param)
6. deepspeed.utils.safe_get_local_grad(param)
7. deepspeed.utils.safe_get_local_optimizer_state(param, optim_state_key)

**Comment by rohan-varma** [comment]
@stas00 Curious if you've tried the `with_grads=True` as @awgu mentioned and if there are any limitations / hiccups you've hit? Also wondering if it would be useful for us to provide APIs similar to deepspeed to reduce complexity (just call a single function over summoning full params, managing memory, and retrieving the grads)? 

The way I see it, we want an iterator allowing users to access tuples `(fqn, unsharded_grad)` for debugging purposes - is this sufficient for your use cases or would there be a better UX possible here?

**Comment by stas00** [comment]
I haven't tried it - I was only asking on slack since I couldn't find how to do that in the docs. And I was requested to repost this as an Issue.

To follow up to your suggestion - won't `summon_full_params()` be wasteful if all the user wants is the grads? It's already expensive, so why add more overhead gathering more data than needed? perhaps the option could be `only_grads=True`? Unless the full params are already there anyway?

The main use would be to track and log grads if the training starts diverging.

**Comment by kiddyboots216** [comment]
I have been using ``with_grads=True`` to implement a gradient compression alg with FSDP and wouldn't report any issues. Is it slow? Oh, definitely. But if you wanted to do this in a non-debug mode, then you would just implement whatever you needed as a hook. I think that tracking and logging gradients does not require ``summon_full_params()`` and could likely be implemented with the existing API, but I'm not sure, I would have to actually do it and see how bad it is.

**Comment by stas00** [comment]
Thank you for your practical feedback, Ashwinee!

If possible let's ask for the fastest version then.

**Comment by awgu** [comment]
Hi all! Thanks for filing this issue.

Adding this support to existing FSDP is non-trivial given the `FlatParameter` construct. (For example, take a look at `_flat_param.py` or `_optim_utils.py` to see some of the logic needed to handle the view-setting and resharding for the flat parameter.)

We have been focusing on an FSDP rewrite that does not have the `FlatParameter` and would help address this ask. See https://github.com/pytorch/pytorch/issues/114299 for the RFC and https://github.com/pytorch/pytorch/issues/120003 for remaining gaps in eager mode.

Under this new per-parameter FSDP, each parameter is represented as a `DTensor` sharded on a 1D mesh with `Shard(0)` placement, and naturally, the optimizer states follow the same representation. Given this, manipulating these sharded parameters, gradients, and optimizer states can directly use `DTensor` APIs, e.g. `full_tensor()` to get the unsharded version.

Some (pseudo)code off the top of my head:
```
from torch.distributed._composable.fsdp import fully_shard  # current private path to FSDP2
lin = nn.Linear(32, 32)
fully_shard(lin)
assert isinstance(lin.weight, DTensor)
assert isinstance(lin.bias, DTensor)

...

unsharded_weight_grad = lin.weight.grad.full_tensor()
assert type(unsharded_weight_grad) is torch.Tensor
```

**Comment by kiddyboots216** [comment]
Hi Andrew,

I don't think that view shaping etc is necessary for FSDP to handle from the FlatParameter. If I think about something like logging the Gradient norm or masking the Gradient, I don't need that -I just need to view the Union of (sharded Gradient on a given rank) + data that's not sharded. 

I appreciate that adding this functionality might be challenging for not much alpha. But realistically stuff like this can be necessary for logging, because if I'm babysitting my run I want to be able to capture these metrics.

**Comment by awgu** [comment]
Hi @kiddyboots216. Thanks for the feedback! Just to check, is your ask a separate one from @stas00 (or are you representing the same ask)?

> I just need to view the Union of (sharded Gradient on a given rank) + data that's not sharded.

1. What is "data that's not sharded" referring to?
2. What is the gap on accessing the sharded gradient on a given rank?

**Comment by kiddyboots216** [comment]
I think my ask is the same. API methods that let me get p.grad.data for full/local, even if that data is a FlatParam and I need to reshape it if necessary for my usecase, would be good. 

As for what is the gap;

Let's say that before I step my optimizer I want to iterate over the named parameters and log the Gradient norms per parameter, or mask out certain parameters (eg freeze a parameter whose Gradient norm is too large). Consider doing this for FSDP with Mistral7B with use_orig_params=True. Some of the params will be output correctly, like q/k/v/o. But some of the params might not be output correctly, like the MLP params, as per this line from the docs;

"An original parameter may have all,
            some, or none of its data present for a given rank. In the none
            case, its data will be like a size-0 empty tensor."

This is the gap between what's currently available (you can do whatever you need to do if that rank owns the entire shard of that param) and what is desired (do various things on any param that has grad.data even if the shard is not owned or only partially owned by that rank). 

As a result you need to summon the full params, but doing this at every step is incredibly slow (in my testing it can slow things down by 5-8x).

Hope this makes sense.

**Comment by awgu** [comment]
@kiddyboots216 Thanks for the explanation! 

> This is the gap between what's currently available (you can do whatever you need to do if that rank owns the entire shard of that param) and what is desired (do various things on any param that has grad.data even if the shard is not owned or only partially owned by that rank).

I wanted to check if I was understanding this correctly. In my head, there are limitations of our flat parameter/current API offering and limitations of sharding parameters/gradients in general.
- For the latter, if gradients are sharded, there is no way to operate on the unsharded gradient without communication. For example, if we want a gradient norm with respect to the unsharded gradient, then we need to all-reduce. This holds regardless of if FSDP shards via flat-parameter sharding or per-parameter sharding.
- For the former, users cannot really rely on what a rank has in its local gradient shard due to the flat-parameter sharding. What is there is non-intuitive and depends on the group of parameters flattened together and the world size. 

What we can offer is this new FSDP we are working on that uses per-parameter sharding. However, that may only help with the former issue, not the latter. Does this make sense?

**Comment by kiddyboots216** [comment]
This makes sense! Thank you for the explanation, I will check out the new version of FSDP and see how I can implement things based on that.

**Comment by awgu** [comment]
@kiddyboots216 The new FSDP has not been released as prototype yet, but we hope to do so this half. The code is there in nightlies, but there are still some gaps (e.g. https://github.com/pytorch/pytorch/issues/120003). We are battle testing it on some internal cases, and a broader announcement should be coming. I mainly wanted to leave some response to mention that we are trying to address these pain points (just in a more long-term / fundamental way 😅 ).

**Comment by stas00** [comment]
@awgu, your [solution](https://github.com/pytorch/pytorch/issues/120879#issuecomment-1977140356) sounds perfect.

Is this new version targetted for torch==2.3?

In order for us users and framework implementors to be able to discussion these things correctly - are you calling it FSDP v2?

**Comment by awgu** [comment]
> Is this new version targetted for torch==2.3?

It looks like the branch cut for torch==2.3 is 3/8 (in four days), which is tight, so I think we should target torch==2.4 instead. 

> In order for us users and framework implementers to be able to discussion these things unambiguously - are you calling this rewrite FSDP v2?

FSDP v2 sounds good or even FSDP2 for brevity.

**Comment by stas00** [comment]
does FSDP depend on pytorch so tightly that it has to follow its schedule and not its own?

**Comment by awgu** [comment]
@stas00 Sorry, I may have misunderstood.

The FSDP2 code is available in nightlies today, and new changes will also be readily available in nightlies. I only mentioned the main PyTorch release in case users wanted to use it in an official release rather than the nightlies.

**Comment by stas00** [comment]
yes, yes, I just mean that when we talk about different FSDP versions ideally those should be independent of pytorch releases, other than requiring the min torch version - but that of course would require splitting off the torch core. But it'd then give you the flexibility of doing your own releases at your own pace.

While FSDP is tightly coupled with pytorch, it's not pytorch per se. But a framework built around it.

It was fine for DDP since it's very static and barely every changing, but FSDP is going to be tweaked and improved a lot, so I'm questioning whether having it part of pytorch core is a good choice for both users and developers of FSDP.

**Comment by awgu** [comment]
These are good points @stas00 . I think this has been discussed some times internally too, and I think in the end, the proposal was still to include it in pytorch repo and stabilize it over time.

Having it exist in pytorch makes integration testing with other components easier like `DTensor`, where we are also adding code changes. cc: @wanchaol 

> p.s. is FSDP pure python?

Yes, the core implementation is pure Python. However, @yifuwang added a new CUDA fast path to `split_with_sizes_copy.out` for copying after all-gather, and @BoyuanFeng will be adding a new aten op for copying before reduce-scatter. Excluding those (so after some future stable release), the FSDP implementation is all Python.

**Comment by stas00** [comment]
Understood. This just makes competition with Deepspeed harder where an important bug fix can get a new release on the same day. And, of course, more difficult usability-wise as it's somewhat drawn in the ocean of Pytorch. Perhaps there can be some other creative solution.

The reason I asked about pure python is that even if FSDP remains as part of the torch core, if it's pure python - there could be a one liner hack that a user could update their whatever installed torch with the bug fix FSDP sub-tree in a pinch, assuming the minimal `torch` requirements have been met. i.e. copy the folder over or pull in the fixed version with PYTHONPATH override.

**Comment by kiddyboots216** [comment]
Someone asked me for the gradient masking thing so I'm just putting it here until FSDPv2 drops. Credit to @rohan-varma. This is useful if you want to freeze, ex, just the MLPs (if you're doing q-v LoRA) as naively doing it in FSDP will not save you any time.

```
def run_optim_step(self):
  for n, p in self.policy.named_parameters():
      if not self.mask_map[n]:
          self.optim_for_param[p].step()
          self.scheduler_for_param[p].step()
          self.optim_for_param[p].zero_grad()

params_to_mask = {} # your own logic, ex: {'mlp'}, make sure you account for all the wrapped_function and stuff
self.mask_map = {}
self.optim_for_param = {}
self.scheduler_for_param = {}
for name, param in self.policy.named_parameters():
    if name in params_to_mask:
        self.mask_map[name] = 1
        print(f"Masking {name}")
    else:
        self.mask_map[name] = 0
        optimizer = getattr(torch.optim, self.config.optimizer)([param], lr=self.config.lr)
        self.optim_for_param[param] = optimizer
        self.scheduler_for_param[param] = torch.optim.lr_scheduler.LambdaLR(optimizer, lr_lambda=lambda step: min(1.0, (step + 1) / (self.config.warmup_steps + 1)))
```
